### ISSUE TITLE
sharing results of korean(+english) bilingual training

### ISSUE THREAD

**Issue by choihk6610** [huggingface/parler-tts#67]
Hello,

Based on your code, I added Korean tokens (using a Korean emotional dataset) to the tokenizer and fine-tuned the model with the LibriTTS R dataset. The Korean dataset is slightly less than 300 hours, similar in size to LibriTTS R but with fewer speakers. 
I did not perform separate emotion classification.

I wanted to share the wandb report, but due to security reasons, I am unable to do so. 
Instead, I am providing the training and evaluation curves as images and some audio files as a compressed file.

Specifically, no preprocessing has been applied to the sentences in korean dataset.
The performance turned out to be better than expected, so I believe that if Korean is included in the pretraining stage, it could yield excellent results in the fine-tuning stage. I hope that Korean will be included in the next version of the model.

Thank you.
<img width="1613" alt="train" src="https://github.com/huggingface/parler-tts/assets/155605706/06b9997c-8f8c-4d5f-b488-ddb331bff499">
<img width="1600" alt="eval" src="https://github.com/huggingface/parler-tts/assets/155605706/469161cf-d066-44ce-9350-431af7cfde13">
[samples.zip](https://github.com/user-attachments/files/15756535/samples.zip)

**Comment by chunping-xt** [comment]
Which tokenizer are you using? If it is flan-t5 then I think it does not support Korean language. I'm also facing the problem that flan-t5 doesn't support my language.
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
tokenizer.decode(tokenizer("어떻게 지내세요?").input_ids)
> '\<unk\> \<unk\>?\</s>'

**Comment by choihk6610** [comment]
@chunping-xt I mentioned that I added Korean tokens to the tokenizer and used them in the text.

**Comment by chunping-xt** [comment]
@choihk6610 I'd appreciate it if you could instruct to add tokens to a language that flan-t5 doesn't support, I'm quite limited in reading codes. Thanks!

**Comment by choihk6610** [comment]
@chunping-xt It might be helpful to refer to the add_special_tokens() function from the following link:
https://huggingface.co/docs/transformers/main_classes/tokenizer

**Comment by chunping-xt** [comment]
@choihk6610, i added tokens to flan-t5, but i got error while training, can you share how you used tokenizer after adding tokens?
#tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
tokenizer = AutoTokenizer.from_pretrained("parler-tts/parler_tts_mini_v0.1")
x_unk = ['new_tok',...]
tokenizer.add_tokens(x_unk) 
tokenizer.save_pretrained('/mnt/f/parler-tts/my_Flan-T5')

and train:
%cd /mnt/f/parler-tts
!accelerate launch ./training/run_parler_tts_training.py \
    --model_name_or_path "parler-tts/parler_tts_mini_v0.1" \
    --feature_extractor_name "parler-tts/dac_44khZ_8kbps" \
    --description_tokenizer_name "/mnt/f/parler-tts/my_Flan-T5" \
    --prompt_tokenizer_name "/mnt/f/parler-tts_v1/my_Flan-T5" \
    ....
    
    and got many lines: ../aten/src/ATen/native/cuda/Indexing.cu:1289: indexSelectLargeIndex: block: [495,0,0], thread: [96,0,0] Assertion `srcIndex < srcSelectDimSize` failed.

**Comment by sanchit-gandhi** [comment]
Awesome results @choihk6610! It's great to hear you were successfully able to fine-tune the base English model for Korean with these steps. The eval loss curve looks great. You can also experiment with a cosine scheduler, which we found to improve performance slightly:

```sh
    --lr_scheduler_stype "cosine" \
```

We shared the samples with our resident Korean speaker, @ArthurZucker, who said:

> ko2 sounds really good in terms of musicality

The next version of Parler (v1) _might_ contain some other high-resource European languages, which should improve multilingual fine-tuning further. If we devise a good data cleaning pipeline, we could also clean [Yodas2](https://huggingface.co/datasets/espnet/yodas2) and train a larger multilingual checkpoint as a follow-up (v2).

**Comment by lyt719** [comment]
> @chunping-xt It might be helpful to refer to the add_special_tokens() function from the following link: https://huggingface.co/docs/transformers/main_classes/tokenizer

May I ask where your Korean vocab comes from? Do you add all vocabulary from other Korean models to Flan t5?

**Comment by chunping-xt** [comment]
@lyt719, I want to finetune with a language other than Korean. The problem is that when changing the vocab size, the model need to resize it, and I don't know how to do it. Anyway I chose another solution and it's been great with parler, with only < 5 hours of training with ~12 hours of data with no prompt field I've got a TTS model, It's surprising how fast finetune. There are 2 wishes I have enough capacity to experiment:
1. Using tokenizer mT5 for multilingual, vocab is much larger than flan-T5, with the same sound length but shorter number of tokens, not sure the training performance is worse than flan-t5, need to experiment to check.
2. Use the encoder/decoder model instead of the current decoder, the reason is that encoder+decoder has much better control, I compared GPT (decoder) and T5/Bart (encoder/decoder) when training NLP, found T5 superior in control capabilities.

**Comment by lyt719** [comment]
> @lyt719, I want to finetune with a language other than Korean. The problem is that when changing the vocab size, the model need to resize it, and I don't know how to do it. Anyway I chose another solution and it's been great with parler, with only < 5 hours of training with ~12 hours of data with no prompt field I've got a TTS model, It's surprising how fast finetune. There are 2 wishes I have enough capacity to experiment:
> 
> 1. Using tokenizer mT5 for multilingual, vocab is much larger than flan-T5, with the same sound length but shorter number of tokens, not sure the training performance is worse than flan-t5, need to experiment to check.
> 2. Use the encoder/decoder model instead of the current decoder, the reason is that encoder+decoder has much better control, I compared GPT (decoder) and T5/Bart (encoder/decoder) when training NLP, found T5 superior in control capabilities.

Thanks! And Do you do these two tricks based on Mini v0.1 finetune or pretrain from scratch?
 I tryied but i can't change tokeizer to any other in finetune, it may generate a lot errors.

**Comment by chunping-xt** [comment]
@lyt719 not yet, it's my dream and waiting for someone to care and implement it. The only thing I did was use the existing flan-t5 tokens to represent my language.

**Comment by lyt719** [comment]
> @lyt719 not yet, it's my dream and waiting for someone to care and implement it. The only thing I did was use the existing flan-t5 tokens to represent my language.

Me too. So how do you solve the language that flan-t5 tokenizer can't encode?

**Comment by chunping-xt** [comment]
@lyt719, eg: suppose your language is all numbers 0-9, and flan-t5 does not support numeric symbols, but supports abc..xyz character symbols, then you do a mapping of numbers to characters, so you have presented your language in characters. A small note is that the training process uses Whisper to calculate word prediction accuracy, fortunately it does not affect the training quality but is only to monitor the visual training results.

**Comment by manigandanp** [comment]
> @choihk6610, i added tokens to flan-t5, but i got error while training, can you share how you used tokenizer after adding tokens? I don't know how to resize_token_embeddings after adding new tokens.
> 
> #tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base") tokenizer = AutoTokenizer.from_pretrained("parler-tts/parler_tts_mini_v0.1") x_unk = ['new_tok',...] tokenizer.add_tokens(x_unk) tokenizer.save_pretrained('/mnt/f/parler-tts/my_Flan-T5')
> 
> and train: %cd /mnt/f/parler-tts !accelerate launch ./training/run_parler_tts_training.py --model_name_or_path "parler-tts/parler_tts_mini_v0.1" --feature_extractor_name "parler-tts/dac_44khZ_8kbps" --description_tokenizer_name "/mnt/f/parler-tts/my_Flan-T5" --prompt_tokenizer_name "/mnt/f/parler-tts_v1/my_Flan-T5" ....
> 
> ```
> and got many lines: ../aten/src/ATen/native/cuda/Indexing.cu:1289: indexSelectLargeIndex: block: [495,0,0], thread: [96,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
> ```

@chunping-xt have you tried resizing the tokenizer?

```model.resize_token_embeddings(len(tokenizer))```

You may have already found a solution, but I'm sharing this for anyone else who is attempting this for the first time. If you've discovered alternative methods for fine-tuning this model on a non-English dataset, please share your strategy and how you accomplished it.

**Comment by SaleemMalikAI** [comment]
Hi i just want to train that model for my local language and i search that the google/flan-t5-base  is also supports the urdu language is it i still need to add an other tokenizer that supports  the urdu language or just run the taining code

**Comment by choihk6610** [comment]
> Hi i just want to train that model for my local language and i search that the google/flan-t5-base is also supports the urdu language is it i still need to add an other tokenizer that supports the urdu language or just run the taining code

Sorry for the delayed response, I didn't realize there were many comments since it's been a while since I last checked in.
Here's how I typically use LoRA in my workflow. 
Note that this is just a code snippet, so you'll need to adapt it to your specific implementation.

Since F5 is only used as a conditioning for embeddings, when using LoRA, you should be able to substitute an embedding model that's appropriate for your language, assuming the embedding vector dimensions/scales are comparable.

Additionally, if description generation is too slow, you can achieve much faster preprocessing by using qwen3.0 (w/o thinking mode) + vllm to infer simple keywords like gender, pitch, speaking speed, and noise levels. 
Since there are already good models on Hugging Face for most of these keywords (except speaking speed), you can easily set up multi-GPU preprocessing pipelines. (I personally use Ray for this.)

```python
import torch
from transformers.utils import logging
from peft import LoraConfig, get_peft_model


def count_parameters(model):
    """Count the number of trainable parameters in the model."""
    size = sum(p.numel() for p in model.parameters() if p.requires_grad) / 1e6
    size = str(int(size)) + "M"
    return size
    

def initialize_new_embeddings(model, original_vocab_size, num_added_tokens):
    """Initialize new embedding tokens using existing embedding statistics."""
    if num_added_tokens > 0:
        with torch.no_grad():
            old_embed_weights = model.model.embed_tokens.weight[:original_vocab_size]
            existing_mean = old_embed_weights.mean(dim=0, keepdim=True)
            existing_std = old_embed_weights.std(dim=0, keepdim=True)

            new_indices = torch.arange(
                original_vocab_size, original_vocab_size + num_added_tokens
            )
            model.model.embed_tokens.weight[new_indices] = torch.normal(
                mean=existing_mean.expand(num_added_tokens, -1),
                std=existing_std.expand(num_added_tokens, -1),
            )


def resize_vocab(model, tokenizer, added_tokens):
    """Resize the model's vocabulary to include speech tokens."""
    original_vocab_size = len(tokenizer)
    num_added_tokens = tokenizer.add_tokens(added_tokens)
    new_token_indices = list(range(original_vocab_size, original_vocab_size + num_added_tokens))
    
    model.resize_token_embeddings(len(tokenizer))
    model.vocab_size = len(tokenizer)

    initialize_new_embeddings(model, original_vocab_size, num_added_tokens)

    # Fixed the function calls and string formatting
    logger.info(f"model size: {count_parameters(model)}")
    logger.info(f"num added token: {num_added_tokens}")
    logger.info(f"Tokenizer vocabulary size: {len(tokenizer)}")
    logger.info(
        f"Model's embedding layer size: {model.model.embed_tokens.weight.size(0)}"
    )
    logger.info(f"Model's lm_head size: {model.lm_head.weight.size(0)}")

    torch.cuda.empty_cache()
    return model, tokenizer, new_token_indices


def get_lora_model(
    model, 
    target_modules=[], 
    new_token_indices=[], 
):
    lora_config = LoraConfig(
        r=8,
        lora_alpha=16,
        target_modules=target_modules,
        lora_dropout=0.01,
        bias="lora_only",
        task_type="CAUSAL_LM",
        trainable_token_indices=new_token_indices, 
    )
    model = get_peft_model(model, lora_config)
    model.print_trainable_parameters()
    return model



added_tokens = your_new_tokens # ["<|TOKEN_1|>", "<|TOKEN_2|>", ...]
target_modules = your_lora_target_modules # ["layer1", "layer2", ...]
model, tokenizer = your_model, your_tokenizer # AutoModel, AutoTokenizer

model, tokenizer, new_token_indices = resize_vocab(model, tokenizer, added_tokens)
model = get_lora_model(model, target_modules, new_token_indices)

# Trainer here
```
